api/client/commands.go

```diff
@@ -139,10 +140,38 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
 		if _, err := os.Stat(root); err != nil {
 			return err
 		}
-		filename := path.Join(root, "Dockerfile")
+
+		absRoot, err := filepath.Abs(root)
+		if err != nil {
+			return err
+		}
+
+		var filename string
+		origDockerfile := *dockerfileName
+
+		if cmd.IsSet("f") || cmd.IsSet("file") {
+			if filename, err = filepath.Abs(*dockerfileName); err != nil {
+				return err
+			}
+		} else {
+			filename = path.Join(absRoot, *dockerfileName)
+		}
```

Comment:
I'm a little confused as to why we've got this extra logic in here detecting whether `-f` or `--file` got passed in.

Couldn't we just change the default to be the empty string, and swap based on that to `CONTEXT/Dockerfile`?  I think it's going to be confusing for users if they see that the default for `-f` is `"Dockerfile"`, but if they explicitly do `-f Dockerfile` it's now actually semantically different, and means the `Dockerfile` in the current directory, regardless of the context.